snapshotter/metrics_integ_test.go

```diff
@@ -0,0 +1,179 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/firecracker-microvm/firecracker-containerd/firecracker-control/client"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/integtest"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/internal/integtest/stargz/fs/source"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	serviceDiscoveryEndpoint = "http://localhost:8080"
+)
+
+func TestSnapshotterMetrics_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+
+	vmID := 0
+
+	testTimeout := 2 * time.Minute
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	ctx = namespaces.WithNamespace(ctx, strconv.Itoa(vmID))
+
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	defer fcClient.StopVM(ctx, &proto.StopVMRequest{VMID: strconv.Itoa(vmID)})
+	require.NoError(t, err, "Failed to create fccontrol client")
+
+	require.NoError(t, pullImageWithRemoteSnapshotterInVM(ctx, strconv.Itoa(vmID), fcClient))
+	verifyMetricsResponse(t, 1)
+}
+
+func TestSnapshotterMetricsMultipleVMs_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+
+	testTimeout := 5 * time.Minute
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	numberOfVms := integtest.NumberOfVms
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	require.NoError(t, err, "Failed to create fccontrol client")
+
+	group, ctx := errgroup.WithContext(ctx)
+
+	for vmID := 0; vmID < numberOfVms; vmID++ {
+		id := vmID
+		ctxNamespace := namespaces.WithNamespace(ctx, strconv.Itoa(id))
+		defer fcClient.StopVM(ctxNamespace, &proto.StopVMRequest{VMID: strconv.Itoa(id)})
+
+		group.Go(
+			func() error {
+				return pullImageWithRemoteSnapshotterInVM(ctxNamespace, strconv.Itoa(id), fcClient)
+			},
+		)
+
+	}
+
+	err = group.Wait()
+	require.NoError(t, err)
+	verifyMetricsResponse(t, numberOfVms)
+}
+
+// TODO (ginglis13): ensure service discovery response changes as VMs are started and stopped.
+// pending https://github.com/firecracker-microvm/firecracker-containerd/issues/651
+// func TestSnapshotterMetricsMultipleVMsStopStart_Isolated(t *testing.T) {}
+
+func pullImageWithRemoteSnapshotterInVM(ctx context.Context, vmID string, fcClient *client.Client) error {
+	client, err := containerd.New(integtest.ContainerdSockPath, containerd.WithDefaultRuntime(integtest.FirecrackerRuntime))
+	if err != nil {
+		return fmt.Errorf("Unable to create client to containerd service at %s, is containerd running?: %v", integtest.ContainerdSockPath, err)
+	}
+	kernelArgs := integtest.DefaultRuntimeConfig.KernelArgs + " 8250.nr_uarts=0 quiet loglevel=1"
```

Comment:
@kzys it hasn't proven to reduce the flakiness but figured it would help increase firecracker vm stability even if we aren't stress testing the in-vm snapshotter.